src/script.rs

```diff
@@ -146,4 +176,61 @@ impl<'a> ScriptInvocation<'a> {
             }
         }
     }
+
+    /// Invokes the script and returns the result.
+    #[inline]
+    pub fn invoke_async<T, C>(&self, con: C) -> RedisFuture<(C, T)>
+    where
+        T: FromRedisValue + Send + 'static,
+        C: AsyncConnectionLike + Send + Clone + 'static,
+    {
+        let code = self.script.inner.code.to_string();
+
+        // Check if the script has already been loaded into Redis by this instance
+        // Note that storing this boolean makes it so that we do not need to clone
+        // the ScriptInvocation in the normal case when we know that the script has
+        // already been loaded.
```

Comment:
Rather than using an atomic to avoid the need clone we could create a custom future for this (as long as we move `self` in this method). Something like this

```rust
let future = Either::Left(EVAL_FUTURE);
let mut did_load = false;
poll_fn(move || loop {
    match &mut future {
        Either::Left(f) => match f.poll() {
            Err(err) => {
                if did_load {
                    return Err(err);
                }
                if error.kind() == NoScriptError {
                    future = Either::Right(LOAD_FUTURE);
                    did_load = true; // Only retry the eval once after loading
                }
            }
        },
        Either::Right(f) => {
           try_ready!(f.poll());
           future = Either::Left(EVAL_FUTURE);
        }
    }
})
```

While it has a different sort of complexity I think I prefer that since it limits it to only the async version.

Reply:
Good point! I implemented it as a custom future. Happy to use `poll_fn` instead if you prefer